design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,216 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+    1. I want to have readable tests for my `State-Machine-Device` a set/return `0x020` is not incredibly readable at the unit test level (although it is needed at the system-test level)
+
+    
+I would then say as part of my proposal that the low-level design (as created by the Plankton team) of stateful devices might better look like this.
+
+```python
+
+class LinkumT95()
+
+   def __init__(self, linkum_t95_machine):
+       if not linkum_t95_machine:
+           self._linkum_t95_machine = LinkumT95Machine()
+       else:
+           self._linkum_t95_machine = linkum_t95_machine
+    
+   def getStatus():
+       temp = self._linkum_t95_machine.getTemp() # Get temp as float
+       # convert this into a hex 
+       # convert this into a hex character array
+       return hex_char_arry
+   
+```
+This will give a better ability to perform the granular testing at the level that we want. 
+
+ 
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
+
+    def run(self):
+
+        while True:
+            start = datetime.now()
+
+            
+            self._adapter.process(delta) # stream adapters would call asyncore.loop(0.1, count=1) as part of this
+
+            
+            count += 1
+            timer += self._delta()
+            self._delta = self._delta()
+            # ....
+```
+
+#### Environment Updates
+
+The separation of the runnable parts of the simulation (now in the `Environment`) from the `Device` and the `Adapter` parts should make updates which affect only the `Environment` easier.
+
+Here's one possibility I've been considering:
+
+`EnvironmentAdapter` sees `Environment` 
+
+```python 
+    class Environment(object):
+
+    def __init__(self, adapter, environ_adapter):
+        self._environ_adapter = environ_adapter
+        self._environ_adapter.set_adaptee(self)
+        self._adapter = adapter
+
+    def run(self):
+
+        while True:
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
```

Comment:
We already have these `import_adapter`-methods and so on, they probably need to be re-written slightly, but basically they are the same. The fact that each device will come with its own set of adapters will make this easier, because we don't need to check "does this combination of device and adapter exist" (we're not doing that at the moment either, everything just fails spectacularly) but only "what adapters does this device come with". One thing we need to think about is naming though. I wouldn't want users to type `-d chopper -p ChopperAdapterEpics`, so we need to tackle this somehow. One option (@owenarnold, you're going to love this) would be to have an `AdapterFactory` where we could subscribe adapters per device. It would look something like this `factory.subscribe(DeviceClass, AdapterClass, 'protocolname')` and then you could create the proper adapter via `factory.createAdapter(device_object, 'protocolname')`.